hbase-asyncfs/src/main/java/org/apache/hadoop/hbase/io/asyncfs/FanOutOneBlockAsyncDFSOutput.java

```diff
@@ -429,6 +429,9 @@ private void flushBuffer(CompletableFuture<Long> future, ByteBuf dataBuf,
       future.completeExceptionally(new IOException("stream already broken"));
       // it's the one we have just pushed or just a no-op
       waitingAckQueue.removeFirst();
+
+      // This method takes ownership of the dataBuf so we need release it before returning.
+      dataBuf.release();
```

Comment:
I guess here we also need to release the headerBuf and checksumBuf?

So maybe we declare these 3 ByteBufs at the top of the method, and use a big try finally to hold the whole method, and in the finally block, we release these 3 ByteBufs? Like this:

```
  private void flushBuffer(CompletableFuture<Long> future, ByteBuf dataBuf,
    long nextPacketOffsetInBlock, boolean syncBlock) {
    ByteBuf checksumBuf = null;
    ByteBuf headerBuf = null;
    try {
      int dataLen = dataBuf.readableBytes();
      int chunkLen = summer.getBytesPerChecksum();
      int trailingPartialChunkLen = dataLen % chunkLen;
      int numChecks = dataLen / chunkLen + (trailingPartialChunkLen != 0 ? 1 : 0);
      int checksumLen = numChecks * summer.getChecksumSize();
      checksumBuf = alloc.directBuffer(checksumLen);
      summer.calculateChunkedSums(dataBuf.nioBuffer(), checksumBuf.nioBuffer(0, checksumLen));
      checksumBuf.writerIndex(checksumLen);
      PacketHeader header = new PacketHeader(4 + checksumLen + dataLen, nextPacketOffsetInBlock,
        nextPacketSeqno, false, dataLen, syncBlock);
      int headerLen = header.getSerializedSize();
      headerBuf = alloc.buffer(headerLen);
      header.putInBuffer(headerBuf.nioBuffer(0, headerLen));
      headerBuf.writerIndex(headerLen);
      Callback c =
        new Callback(future, nextPacketOffsetInBlock + dataLen, datanodeInfoMap.keySet(), dataLen);
      waitingAckQueue.addLast(c);
      // recheck again after we pushed the callback to queue
      if (state != State.STREAMING && waitingAckQueue.peekFirst() == c) {
        future.completeExceptionally(new IOException("stream already broken"));
        // it's the one we have just pushed or just a no-op
        waitingAckQueue.removeFirst();
        return;
      }
      // TODO: we should perhaps measure time taken per DN here;
      // we could collect statistics per DN, and/or exclude bad nodes in createOutput.
      for (Channel ch : datanodeInfoMap.keySet()) {
        ch.write(headerBuf.retainedDuplicate());
        ch.write(checksumBuf.retainedDuplicate());
        ch.writeAndFlush(dataBuf.retainedDuplicate());
      }
      nextPacketSeqno++;
    } finally {
      ReferenceCountUtil.safeRelease(checksumBuf);
      ReferenceCountUtil.safeRelease(headerBuf);
      ReferenceCountUtil.safeRelease(dataBuf);
    }
  }
```

Reply:
You are right... that said I think just do the release explicit in the if block is cleaner. Also `safeRelease` is usually a code smell as it means you might call release too often.

Reply:
(I added the explicit release calls)

Reply:
Ah, yeah. Just call ByteBuf.release directly is also fine but we need to do the null check by ourselves.